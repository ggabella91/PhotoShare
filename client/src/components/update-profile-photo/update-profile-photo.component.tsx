import React, { useState, useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { FieldsToUpdate } from '../../redux/user/user.types';
import { changeInfoStart } from '../../redux/user/user.actions';

import {
  updateProfilePhotoStart,
  clearProfilePhotoStatuses,
} from '../../redux/post/post.actions';
import { AppState } from '../../redux/root-reducer';

import { FormFileInput } from '../form-input/form-input.component';
import Button from '../button/button.component';

import Alert from 'react-bootstrap/Alert';

import './update-profile-photo.styles.scss';

interface ImgPreview {
  src: string;
  alt: string;
}

export const UpdateProfilePhoto: React.FC = () => {
  const [profilePhoto, setProfilePhoto] = useState<FormData | null>(null);
  const [imgPreview, setImgPreview] = useState<ImgPreview | null>(null);
  const [fileInputKey, setFileInputKey] = useState(1609996842790);
  const [profilePhotoStatus, setProfilePhotoStatus] = useState({
    success: false,
    error: false,
  });
  const [showProfilePhotoAlert, setShowProfilePhotoAlert] = useState(false);
  const fileInputRef = useRef<HTMLInputElement | null>(null);

  const userState = useSelector((state: AppState) => state.user);
  const postState = useSelector((state: AppState) => state.post);
  const dispatch = useDispatch();

  const { currentUser } = userState;
  const {
    profilePhotoConfirm: updateProfilePhotoConfirm,
    profilePhotoError: updateProfilePhotoError,
    profilePhotoKey,
  } = postState;

  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files?.length) {
      const file = event.target.files[0];

      const formData = new FormData();

      formData.append('profile-photo', file, file.name);

      setProfilePhoto(formData);
      setImgPreview({ src: URL.createObjectURL(file), alt: file.name });
    } else {
      setProfilePhoto(null);
      setImgPreview(null);
    }
  };

  const handleSubmitProfilePhoto = async (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    if (currentUser) {
      event.preventDefault();
      setProfilePhotoStatus({ success: false, error: false });

      if (profilePhoto) {
        setShowProfilePhotoAlert(true);

        dispatch(updateProfilePhotoStart(profilePhoto));
        setTimeout(() => setShowProfilePhotoAlert(false), 5000);
      }

      setFileInputKey(Date.now());

      setProfilePhoto(null);
      setImgPreview(null);
    }
  };

  useEffect(() => {
    if (currentUser) {
      if (updateProfilePhotoError) {
        setProfilePhotoStatus({ ...profilePhotoStatus, error: true });
      } else if (updateProfilePhotoConfirm) {
        setProfilePhotoStatus({ ...profilePhotoStatus, success: true });
        if (profilePhotoKey && currentUser) {
          const fieldsToUpdate: FieldsToUpdate = {
            name: currentUser.name,
            email: currentUser.email,
            username: currentUser.username,
            bio: currentUser.bio,
            photo: profilePhotoKey,
          };

          dispatch(changeInfoStart(fieldsToUpdate));
        }
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [updateProfilePhotoError, updateProfilePhotoConfirm]);

  const handleRenderAlert = (type: string, message: string) => {
    if (currentUser) {
      dispatch(clearProfilePhotoStatuses());
      setTimeout(() => {
        setProfilePhotoStatus({ success: false, error: false });
        setShowProfilePhotoAlert(false);
      }, 3000);
      return (
        <Alert
          variant={type}
          className='photo-alert'
          onClose={handleHideProfilePhotoAlert}
          dismissible
        >
          {message}
        </Alert>
      );
    }
  };

  const handleHideProfilePhotoAlert = () => setShowProfilePhotoAlert(false);

  return (
    <div className='settings'>
      <div className='update-profile-photo'>
        <span>Update your profile photo</span>
        <div className='profile-photo-container'>
          {imgPreview || showProfilePhotoAlert ? null : (
            <div className='img-preview-placeholder'>
              <div className='placeholder-text-container'>
                <span className='placeholder-text'>
                  Select a photo to preview it here
                </span>
              </div>
            </div>
          )}
          {!imgPreview && showProfilePhotoAlert ? (
            <div className='settings-alert'>
              {profilePhotoStatus.error
                ? handleRenderAlert('danger', 'Error updating profile picture.')
                : null}
              {profilePhotoStatus.success
                ? handleRenderAlert(
                    'success',
                    'Profile picture changed successfully!'
                  )
                : null}
            </div>
          ) : null}
          {imgPreview ? (
            <img
              className='img-preview'
              src={imgPreview ? imgPreview.src : ''}
              alt={imgPreview ? imgPreview.alt : ''}
            />
          ) : null}
        </div>
        <form encType='multipart/form-data' onSubmit={handleSubmitProfilePhoto}>
          <FormFileInput
            name='profile-photo'
            type='file'
            label='Select photo'
            accept='image/*'
            onChange={handleFileChange}
            key={fileInputKey}
            fileName={fileInputRef.current?.files?.[0]?.name || ''}
            inputRef={fileInputRef}
          />
          <div className='button'>
            <Button
              className='submit-button'
              onClick={handleSubmitProfilePhoto}
            >
              Upload photo
            </Button>
          </div>
        </form>
      </div>
    </div>
  );
};

export default UpdateProfilePhoto;
