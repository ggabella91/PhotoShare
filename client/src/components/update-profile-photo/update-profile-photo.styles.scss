.update-profile-photo {
  position: relative;
  margin: 20px 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  .profile-photo-container {
    margin-top: 15px;
    width: 310px;
    height: 310px;
    border: solid 2px;
    position: relative;

    .placeholder-text-container {
      position: relative;
      margin-top: 50%;
      margin-left: 50%;
      transform: translate(-50%, -50%);

      .placeholder-text {
        font-weight: 420;
      }
    }

    .img-preview {
      margin-top: 50%;
      transform: translateY(-50%);
      position: relative;
      max-width: 300px;
      max-height: 300px;
      height: auto;
    }

    .photo-alert {
      position: relative;
      max-width: 260px;
      margin-top: 50%;
      margin-left: 50%;
      transform: translate(-50%, -50%);
    }
  }

  span {
    font-weight: 400;
    margin: 0px;
    width: 300px;
    line-height: 1.5;
  }
}
