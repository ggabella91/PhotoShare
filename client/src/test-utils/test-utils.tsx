import { FC, ReactElement } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render, RenderOptions } from '@testing-library/react';
import { store } from '../redux/store';

interface WrapperProps {
  wrapperProps?: { route: string; location: string };
}

type CustomRenderOptions = RenderOptions & WrapperProps;

const WithReduxProvider: FC<WrapperProps> = ({ children, wrapperProps }) => {
  return (
    <Provider store={store}>
      <BrowserRouter>
        <Routes>
          <Route path={wrapperProps?.route || '/'} element={children} />
        </Routes>
      </BrowserRouter>
    </Provider>
  );
};

const renderWithReduxAndRouter = (
  ui: ReactElement,
  options?: Omit<CustomRenderOptions, 'wrapper'>
) => {
  window.history.pushState({}, '', options?.wrapperProps?.location || '/');
  const customContainer = document.createElement('div');
  customContainer.id = 'root';

  return render(ui, {
    wrapper: (props) => (
      <WithReduxProvider {...props} wrapperProps={options?.wrapperProps} />
    ),
    container: document.body.appendChild(customContainer),
    ...options,
  });
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { renderWithReduxAndRouter as render };
