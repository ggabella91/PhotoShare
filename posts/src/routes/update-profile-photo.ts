import express, { Request, Response } from 'express';
import multer from 'multer';
import { resizePhoto } from '../utils/resize';
import { Post } from '../models/post';
import { requireAuth, BadRequestError } from '@ggabella-photo-share/common';
import { buffToStream } from '../utils/buffToStream';
import { AWS } from '../index';
import { S3 } from 'aws-sdk';

const router = express.Router();

const fileStorage = multer.memoryStorage();

const imageFilter = (
  req: Request,
  file: Express.Multer.File,
  cb: multer.FileFilterCallback
): void => {
  if (file.mimetype.startsWith('image')) {
    cb(null, true);
  } else {
    cb(new BadRequestError('Only images can be uploaded'));
  }
};

const upload = multer({ storage: fileStorage, fileFilter: imageFilter });

router.post(
  '/api/posts',
  requireAuth,
  upload.single('photo'),
  resizePhoto,
  async (req: Request, res: Response) => {
    const caption = req.body.caption || '';

    const key = req.file.filename;

    const s3 = new AWS.S3();

    const uploadParams: S3.Types.PutObjectRequest = {
      Bucket: 'photo-share-app-profile-photos',
      Key: key,
      Body: '',
    };

    const fileBuffer = req.file.buffer;
    const fileStream = buffToStream(fileBuffer);

    uploadParams.Body = fileStream;

    let location = '';

    s3.upload(uploadParams, async (err, data) => {
      if (err) {
        console.log(err);
        throw new Error('Error uploading the photo');
      }
      if (data) {
        location = data.Location;
        console.log('Upload success!', location);

        // const post = Post.build({
        //   fileName: req.file.originalname,
        //   caption,
        //   createdAt: new Date(),
        //   userId: req.currentUser!.id,
        //   s3Key: key,
        //   s3ObjectURL: location,
        // });

        // await post.save();

        // res.status(201).send(post);
      }
    });
  }
);

export { router as createPostRouter };
