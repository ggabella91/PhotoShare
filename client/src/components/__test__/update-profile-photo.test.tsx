import { render, screen, userEvent } from '../../test-utils/test-utils';

import { UpdateProfilePhoto } from '../update-profile-photo/update-profile-photo.component';

describe('update-profile-photo component tests', () => {
  const setup = () => {
    global.URL.createObjectURL = jest.fn();

    render(<UpdateProfilePhoto />);
  };

  it('renders an update-profile-photo component', () => {
    setup();

    const updateProfilePhoto = screen.getByText(/Update your profile photo/i);

    expect(updateProfilePhoto).toBeInTheDocument();
  });

  it('uploading a photo and clicking clicking upload photo calls update profile photo action creator', () => {
    setup();

    const testPhotoFile = new File(['test-photo-file'], 'test-photo', {
      type: 'img/jpeg',
    });

    const fileInput: HTMLInputElement = screen.getByTestId('file-input');

    const uploadButton = screen.getByText(/Upload photo/i);

    userEvent.upload(fileInput, testPhotoFile);

    userEvent.click(uploadButton);

    if (fileInput.files) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(fileInput.files[0]).toStrictEqual(testPhotoFile);
    }
    expect(global.URL.createObjectURL).toBeCalled();
  });
});
