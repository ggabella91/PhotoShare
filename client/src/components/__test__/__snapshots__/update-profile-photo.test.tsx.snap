// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an update-password component 1`] = `
<div
  className="settings"
>
  <div
    className="update-profile-photo"
  >
    <span>
      Update your profile photo
    </span>
    <div
      className="profile-photo-container"
    >
      <div
        className="img-preview-placeholder"
      >
        <div
          className="placeholder-text-container"
        >
          <span
            className="placeholder-text"
          >
            Select a photo to preview it here
          </span>
        </div>
      </div>
    </div>
    <form
      encType="multipart/form-data"
      onSubmit={[Function]}
    >
      <FormFileInput
        accept="image/*"
        key="1609996842790"
        label="Select photo"
        name="profile-photo"
        onChange={[Function]}
        type="file"
      />
      <div
        className="button"
      >
        <Button
          className="submit-button"
          onClick={[Function]}
        >
          Upload photo
        </Button>
      </div>
    </form>
  </div>
</div>
`;
